File: packages/xyd-opensdk-ruby/src/rbtype.ts
import type { TypeRef } from '@xyd-js/opensdk-core';

import { pascalCase } from './naming';

/**
 * Map an IR TypeRef to a YARD documentation type string (used only in `@param`
 * / `@return` comments — Ruby is duck-typed, so the emitted code never depends
 * on these). Named refs render under the `Models::` namespace so the doc links
 * resolve, scalars map to Ruby builtins, arrays/maps recurse.
 */
export function rbDocType(ref: TypeRef | undefined): string {
  if (!ref) return 'Object';
  switch (ref.kind) {
    case 'scalar':
      return rbScalar(ref.scalar, ref.format);
    case 'ref':
      return ref.name ? `Models::${pascalCase(ref.name)}` : 'Object';
    case 'array':
      return `Array<${rbDocType(ref.items as TypeRef | undefined)}>`;
    case 'map':
      return `Hash{String => ${rbDocType(ref.values as TypeRef | undefined)}}`;
    default:
      return 'Object';
  }
}

function rbScalar(scalar: string | undefined, format: string | undefined): string {
  switch (scalar) {
    case 'string':
      return format === 'binary' ? 'IO, String' : 'String';
    case 'integer':
      return 'Integer';
    case 'number':
      return 'Float';
    case 'boolean':
      return 'Boolean';
    default:
      return 'Object';
  }
}

File: crates/xyd_opensdk_ruby/src/rbtype.rs
//! Port of rbtype.ts — IR TypeRef → YARD doc-type string (used only in
//! `@param`/`@return` comments).

use serde_json::Value;

use crate::naming::pascal_case;

pub fn rb_doc_type(ref_: Option<&Value>) -> String {
    let Some(r) = ref_ else {
        return "Object".to_string();
    };
    match r.get("kind").and_then(|k| k.as_str()) {
        Some("scalar") => rb_scalar(
            r.get("scalar").and_then(|s| s.as_str()),
            r.get("format").and_then(|f| f.as_str()),
        ),
        Some("ref") => match r.get("name").and_then(|n| n.as_str()) {
            Some(n) if !n.is_empty() => format!("Models::{}", pascal_case(n)),
            _ => "Object".to_string(),
        },
        Some("array") => format!("Array<{}>", rb_doc_type(r.get("items"))),
        Some("map") => format!("Hash{{String => {}}}", rb_doc_type(r.get("values"))),
        _ => "Object".to_string(),
    }
}

fn rb_scalar(scalar: Option<&str>, format: Option<&str>) -> String {
    match scalar {
        Some("string") => {
            if format == Some("binary") {
                "IO, String".to_string()
            } else {
                "String".to_string()
            }
        }
        Some("integer") => "Integer".to_string(),
        Some("number") => "Float".to_string(),
        Some("boolean") => "Boolean".to_string(),
        _ => "Object".to_string(),
    }
}
